projects/igniteui-angular/src/lib/grids/grid-base.component.ts

```diff
@@ -4647,113 +4677,18 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
     /**
      * @hidden
      */
-    public onHeaderCheckboxClick(event, filteredData) {
-        this.allRowsSelected = event.checked;
-        const newSelection =
-            event.checked ?
-                filteredData ?
-                    this.selection.add_items(this.id, this.selection.get_all_ids(filteredData, this.primaryKey)) :
-                    this.selection.get_all_ids(this.gridAPI.get_all_data(true), this.primaryKey) :
-                filteredData ?
-                    this.selection.delete_items(this.id, this.selection.get_all_ids(filteredData, this.primaryKey)) :
-                    this.selection.get_empty();
-        this.triggerRowSelectionChange(newSelection, null, event, event.checked);
-        this.checkHeaderCheckboxStatus(event.checked);
+    public onHeaderSelectorClick(event) {
+        this.selectionService.areAllRowSelected() ?
```

Comment:
```areAllRowSelected``` is going to be slow in the worst case scenario. At some point it tries to create a Set of all data rows by their IDs (could be thousands right?) . Isn't it possible to get the information from the checkbox state instead or otherwise cache this?

Reply:
Currently we use the checkbox only for presentation of the state and we prevent default checkbox behavior  on click, because the whole row selector aria trigger select/deselect all behavior. If you check areAllRowSelected method, you can see that the state is cached and recalculated only there is actual need.